Connectors/Rhino/Speckle.Connectors.GrasshopperShared/Operations/Send/GrasshopperBlockPacker.cs

```diff
@@ -22,42 +26,81 @@ internal sealed class GrasshopperBlockPacker
   public Dictionary<string, InstanceDefinitionProxy> InstanceDefinitionProxies { get; } = [];
 
   /// <summary>
-  /// Processes a <see cref="SpeckleBlockInstanceWrapper"/> by validating inputs before delegating to private method.
-  /// Returns the <see cref="SpeckleObjectWrapper"/>s that make up the definition(s). These need to be added to the collection.
+  /// Processes a <see cref="SpeckleBlockInstanceWrapper"/> by tracking it in InstanceObjectsManager and recursively
+  /// processing its definition. Handles depth calculation for nested block hierarchies.
   /// </summary>
-  public List<SpeckleObjectWrapper>? ProcessInstance(SpeckleBlockInstanceWrapper? blockInstance)
+  /// <param name="blockInstance">The block instance to process</param>
+  /// <param name="depth">Current nesting depth (0 = top level, increases for nested instances)</param>
+  public List<SpeckleObjectWrapper>? ProcessInstance(SpeckleBlockInstanceWrapper? blockInstance, int depth = 0)
   {
-    // NOTE: in pure gh workflows Definition might be null ALTHOUGH we're throwing an Exception 🤨. But rather safe than sorry.
-    if (blockInstance?.Definition != null)
+    if (blockInstance?.Definition == null)
     {
-      return ProcessDefinition(blockInstance.Definition);
+      return null;
     }
 
-    return null;
+    var instanceId = blockInstance.ApplicationId ?? Guid.NewGuid().ToString(); // Safety net - final validation before object tracking
```

Comment:
if this appId happens to be null and you general a new one, assign it to the instance wrapper (since this will implicitly assign it to the wrapped instance also). Like what you are doing below with definitions

Reply:
Fixed :) using null-coalescing assignment on **wrapper's** `ApplicationId`

Reply:
[fix: ApplicationId assignment consistency across block wrappers](https://github.com/specklesystems/speckle-sharp-connectors/pull/934/commits/2f2ee8204c6704169cd56615c011c05b25d5cb28)